File: linux-mainline/scripts/dagu-touch-boost.py
#!/usr/bin/env python3
"""Raise GPU/CPU floors while touching, Chrome video, or camera SoftISP.

Idle stays on simple_ondemand / schedutil (GPU min 587 MHz). A hold-drag
or a <video> used to bounce 587↔670 and leave the prime core at 845 MHz,
which is the felt hitch / dropped frame. This does not change Chrome
CPU vs GPU raster. Do not set governor=performance.

Himax is spi-gpio bitbang (IRQF_ONESHOT thread). SoftISP on 5MP/12MP RAW
saturates the cluster and mutter's libinput loop starves — taps still
wake the backlight via logind, but folders/close-window do not respond.
Do not pin SoftISP to the big cluster: pin it to silver CPU0-3 and keep
Himax/mutter on Gold. Signal is any userspace fd on `/dev/video0` or
`/dev/video3` (plus gst-launch holding the loopback nodes).
"""
from __future__ import annotations

import glob
import os
import struct
import time

EV_KEY, EV_ABS = 0x01, 0x03
ABS_MT_TRACKING_ID = 0x39
BTN_TOUCH = 0x14A
EVENT = struct.Struct("llHHi")

GPU_MIN = "/sys/class/devfreq/3d00000.gpu/min_freq"
GPU_IDLE = "587000000"
GPU_HOLD = "670000000"

CPU = (
    ("/sys/devices/system/cpu/cpufreq/policy0/scaling_min_freq", "300000", "1248000"),
    ("/sys/devices/system/cpu/cpufreq/policy4/scaling_min_freq", "710400", "1766400"),
    ("/sys/devices/system/cpu/cpufreq/policy7/scaling_min_freq", "844800", "1977600"),
)

HOLD_TAIL_S = 0.18
VIDEO_SCAN_S = 0.12
SOFTISP_CPUS = frozenset({0, 1, 2, 3})
ALL_CPUS = frozenset(range(os.cpu_count() or 8))
# comm is TASK_COMM_LEN=16 including NUL → 15 chars.
SKIP_PIN = {
    "gnome-shell",
    "Xwayland",
    "dagu-touch-boo",
    "dagu-himax-irq",
}


def write(path: str, value: str) -> None:
    try:
        with open(path, "w") as f:
            f.write(value + "\n")
    except OSError:
        pass


def find_himax() -> str:
    for name in sorted(glob.glob("/sys/class/input/event*/device/name")):
        try:
            text = open(name, "r", encoding="utf-8", errors="ignore").read()
        except OSError:
            continue
        if "Himax" in text or "HX83121" in text:
            return "/dev/input/" + name.split("/")[4]
    return "/dev/input/event3"


def apply(hold: bool) -> None:
    write(GPU_MIN, GPU_HOLD if hold else GPU_IDLE)
    for path, idle, boosted in CPU:
        write(path, boosted if hold else idle)


BROWSER_MARKERS = (
    b"/opt/google/chrome/chrome",
    b"/usr/lib/chromium/chromium",
    b"/usr/bin/chromium",
)
CAM_NODES = ("/dev/video0", "/dev/video3")
LOOP_NODES = ("/dev/video20", "/dev/video21")
CAM_COMMS = {
    "snapshot",
    "gnome-snapshot",
    "pipewire",
    "wireplumber",
    "cam",
    "gst-launch-1.0",
}
VENUS_NODES = (
    "/dev/video14",
    "/dev/video15",
    "/dev/video-dec0",
    "/dev/video-enc0",
)


def _is_browser(cmd: bytes) -> bool:
    return any(m in cmd for m in BROWSER_MARKERS)


def _comm(pid: str) -> str:
    try:
        with open(f"/proc/{pid}/comm", "r", encoding="utf-8", errors="ignore") as f:
            return f.read().strip()
    except OSError:
        return ""


def _is_cam_comm(comm: str) -> bool:
    if comm in CAM_COMMS:
        return True
    return comm.startswith("gst-launch") or comm.startswith("gnome-snapsho")


def _has_fd(pid: str, nodes: tuple[str, ...]) -> bool:
    try:
        fds = os.listdir(f"/proc/{pid}/fd")
    except OSError:
        return False
    for fd in fds:
        try:
            target = os.readlink(f"/proc/{pid}/fd/{fd}")
        except OSError:
            continue
        if target in nodes:
            return True
    return False


def _has_venus_fd(pid: str) -> bool:
    return _has_fd(pid, VENUS_NODES)


def chrome_video_playing() -> bool:
    """True while Chromium/Chrome is in a Venus or video-compositor path."""
    try:
        pids = os.listdir("/proc")
    except OSError:
        return False
    for pid in pids:
        if not pid.isdigit():
            continue
        try:
            with open(f"/proc/{pid}/cmdline", "rb") as f:
                cmd = f.read()
        except OSError:
            continue
        if not _is_browser(cmd):
            continue
        if _has_venus_fd(pid):
            return True
        if b"--type=" in cmd and b"--type=renderer" not in cmd:
            continue
        task = f"/proc/{pid}/task"
        try:
            tids = os.listdir(task)
        except OSError:
            continue
        for tid in tids:
            try:
                with open(f"{task}/{tid}/comm", "r", encoding="utf-8", errors="ignore") as f:
                    name = f.read().strip()
            except OSError:
                continue
            if name.startswith(("VideoFrame", "FFmpeg", "VideoDecod", "V4L2", "Media")):
                return True
    return False


def _iter_pids():
    try:
        pids = os.listdir("/proc")
    except OSError:
        return
    for pid in pids:
        if pid.isdigit():
            yield pid


def _has_softisp_thread(pid: str) -> bool:
    task = f"/proc/{pid}/task"
    try:
        tids = os.listdir(task)
    except OSError:
        return False
    for tid in tids:
        try:
            with open(f"{task}/{tid}/comm", "r", encoding="utf-8", errors="ignore") as f:
                name = f.read().strip()
        except OSError:
            continue
        if name.startswith(("SWIspWorker", "DebayerCpu")):
            return True
    return False


def camera_streaming() -> bool:
    """True while Snapshot/cam/gst is up, or a SoftISP worker thread exists.

    pipewire/wireplumber keep `/dev/video0` open for enumeration. That is
    not STREAMON — do not boost or pin from the fd alone.
    """
    for pid in _iter_pids():
        comm = _comm(pid)
        if comm in ("snapshot", "gnome-snapshot", "cam") or comm.startswith(
            ("gst-launch", "gnome-snapsho")
        ):
            return True
        if _has_softisp_thread(pid):
            return True
        if _is_cam_comm(comm) and _has_fd(pid, LOOP_NODES):
            return True
    return False


def pin_softisp(enable: bool) -> None:
    """Keep DebayerCpu/libcamera off Gold so mutter and Himax still run."""
    cpus = SOFTISP_CPUS if enable else ALL_CPUS
    for pid in _iter_pids():
        comm = _comm(pid)
        if comm in SKIP_PIN:
            continue
        if not (_has_fd(pid, CAM_NODES) or _has_fd(pid, LOOP_NODES)):
            continue
        try:
            os.sched_setaffinity(int(pid), cpus)
        except (OSError, ValueError, PermissionError):
            continue


def main() -> None:
    dev = find_himax()
    fd = os.open(dev, os.O_RDONLY | os.O_NONBLOCK)
    apply(False)
    pin_softisp(False)
    contacts = 0
    holding = False
    drop_at = None
    video = False
    last_scan = 0.0
    pinned = False
    while True:
        now = time.monotonic()
        try:
            buf = os.read(fd, EVENT.size * 64)
        except BlockingIOError:
            buf = b""
        if buf:
            for off in range(0, len(buf) - EVENT.size + 1, EVENT.size):
                _sec, _usec, typ, code, value = EVENT.unpack_from(buf, off)
                if typ == EV_ABS and code == ABS_MT_TRACKING_ID:
                    if value >= 0:
                        contacts += 1
                    elif contacts > 0:
                        contacts -= 1
                elif typ == EV_KEY and code == BTN_TOUCH:
                    contacts = 1 if value else 0
        if now - last_scan >= VIDEO_SCAN_S:
            video = chrome_video_playing() or camera_streaming()
            if video and not pinned:
                pin_softisp(True)
                pinned = True
            elif not video and pinned:
                pin_softisp(False)
                pinned = False
            last_scan = now
        want = contacts > 0 or video
        if want:
            drop_at = None
            if not holding:
                holding = True
                apply(True)
        elif holding:
            if drop_at is None:
                drop_at = now + HOLD_TAIL_S
            elif now >= drop_at:
                holding = False
                drop_at = None
                apply(False)
        time.sleep(0.004)


if __name__ == "__main__":
    main()

File: linux-mainline/userspace/dagu-touch-boost.c
/*
 * dagu: raise GPU/CPU floors while Himax reports contact.
 *
 * Replaces dagu-touch-boost.py. Does not walk /proc fd tables (that froze
 * mutter). SoftISP pinning lives in dagu_cam.cpp. Do not set
 * governor=performance. Kernel himax-dagu also votes the same floors; double
 * apply is harmless.
 */
#define _GNU_SOURCE
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <linux/input.h>
#include <poll.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define GPU_MIN "/sys/class/devfreq/3d00000.gpu/min_freq"
#define GPU_IDLE "587000000"
#define GPU_HOLD "670000000"
#define HOLD_TAIL_NS 180000000L

static const struct {
	const char *path;
	const char *idle;
	const char *hold;
} cpu[] = {
	{ "/sys/devices/system/cpu/cpufreq/policy0/scaling_min_freq", "300000", "1248000" },
	{ "/sys/devices/system/cpu/cpufreq/policy4/scaling_min_freq", "710400", "1766400" },
	{ "/sys/devices/system/cpu/cpufreq/policy7/scaling_min_freq", "844800", "1977600" },
};

static void wr(const char *path, const char *val)
{
	int fd = open(path, O_WRONLY | O_CLOEXEC);

	if (fd < 0)
		return;
	if (write(fd, val, strlen(val)) < 0) {
		/* ignore */
	}
	close(fd);
}

static void apply(int hold)
{
	wr(GPU_MIN, hold ? GPU_HOLD : GPU_IDLE);
	for (size_t i = 0; i < sizeof(cpu) / sizeof(cpu[0]); i++)
		wr(cpu[i].path, hold ? cpu[i].hold : cpu[i].idle);
}

static int find_himax(char *out, size_t n)
{
	DIR *d = opendir("/sys/class/input");
	struct dirent *ent;

	if (!d)
		return -1;
	while ((ent = readdir(d))) {
		char path[128], name[128];
		int fd, len;

		if (strncmp(ent->d_name, "event", 5) != 0 ||
		    strlen(ent->d_name) > 16)
			continue;
		snprintf(path, sizeof(path),
			 "/sys/class/input/%s/device/name", ent->d_name);
		fd = open(path, O_RDONLY | O_CLOEXEC);
		if (fd < 0)
			continue;
		len = read(fd, name, sizeof(name) - 1);
		close(fd);
		if (len <= 0)
			continue;
		name[len] = 0;
		if (!strstr(name, "Himax") && !strstr(name, "HX83121"))
			continue;
		snprintf(out, n, "/dev/input/%s", ent->d_name);
		closedir(d);
		return 0;
	}
	closedir(d);
	snprintf(out, n, "/dev/input/event3");
	return 0;
}

static long long now_ns(void)
{
	struct timespec ts;

	clock_gettime(CLOCK_MONOTONIC, &ts);
	return (long long)ts.tv_sec * 1000000000LL + ts.tv_nsec;
}

int main(void)
{
	char dev[64];
	int fd, contacts = 0, holding = 0;
	long long drop_at = 0;
	struct input_event ev[64];

	find_himax(dev, sizeof(dev));
	for (int i = 0; i < 60; i++) {
		fd = open(dev, O_RDONLY | O_CLOEXEC);
		if (fd >= 0)
			break;
		find_himax(dev, sizeof(dev));
		sleep(1);
	}
	if (fd < 0) {
		fprintf(stderr, "dagu-touch-boost: cannot open %s: %s\n",
			dev, strerror(errno));
		return 1;
	}
	apply(0);
	for (;;) {
		struct pollfd pfd = { .fd = fd, .events = POLLIN };
		int pr = poll(&pfd, 1, 4);
		long long t = now_ns();

		if (pr > 0 && (pfd.revents & POLLIN)) {
			ssize_t n = read(fd, ev, sizeof(ev));
			size_t k, n_ev;

			if (n > 0) {
				n_ev = (size_t)n / sizeof(ev[0]);
				for (k = 0; k < n_ev; k++) {
					if (ev[k].type == EV_ABS &&
					    ev[k].code == ABS_MT_TRACKING_ID) {
						if (ev[k].value >= 0)
							contacts++;
						else if (contacts > 0)
							contacts--;
					} else if (ev[k].type == EV_KEY &&
						   ev[k].code == BTN_TOUCH) {
						contacts = ev[k].value ? 1 : 0;
					}
				}
			}
		}
		if (contacts > 0) {
			drop_at = 0;
			if (!holding) {
				holding = 1;
				apply(1);
			}
		} else if (holding) {
			if (!drop_at)
				drop_at = t + HOLD_TAIL_NS;
			else if (t >= drop_at) {
				holding = 0;
				drop_at = 0;
				apply(0);
			}
		}
	}
}
